engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java

```diff
@@ -86,13 +90,26 @@ public void initialize() {
     PackageManager<TruffleFile> packageManager = new PackageManager<>(fs);
 
     List<TruffleFile> packagePaths = OptionsHelper.getPackagesPaths(environment);
-    packages.addAll(
-        packagePaths.stream()
-            .map(packageManager::fromDirectory)
-            .map(ScalaConversions::asJava)
-            .filter(Optional::isPresent)
-            .map(Optional::get)
-            .collect(Collectors.toList()));
+
+    // Add user packages one-by-one, shadowing previously added packages. It assumes that the
+    // standard library packages will not clash. In the future, we should be able to disambiguate
+    // packages that clash.
+    for (var packagePath : packagePaths) {
+      var asPackage = ScalaConversions.asJava(packageManager.fromDirectory(packagePath));
+      if (asPackage.isPresent()) {
+        var pkg = asPackage.get();
+        var nameExists =
+            packages.stream().filter(p -> p.config().name().equals(pkg.name())).findFirst();
+        nameExists.ifPresent(
+            truffleFilePackage -> {
+              shadowedPackages.add(
+                  new ShadowedPackage(
+                      truffleFilePackage.root().getPath(), pkg.root().getPath(), pkg.name()));
+              packages.remove(truffleFilePackage);
```

Comment:
Shouldn't some HashMap be used here?

Currently, the `filter` and `remove` will result in O(N^2) performance and I'd imagine that the users will be able to have lots of packages installed.